ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactor.java

```diff
@@ -50,73 +45,92 @@ private RestoreCommandsCompactor() {
-    final List<QueuedCommand> compacted = new LinkedList<>(restoreCommands);
-
-    final Set<QueuedCommand> terminatedQueries =
-        findTerminatedQueriesAndRemoveTerminateCommands(compacted);
+    final Map<QueryId, CompactedNode> latestNodeWithId = new HashMap<>();
+    CompactedNode current = null;
+
+    for (final QueuedCommand cmd : restoreCommands) {
+      // Whenever a new command is processed, we check if a previous command with
+      // the same queryID exists - in which case, we mark that command as "shouldSkip"
+      // and it will not be included in the output
+      current = CompactedNode.of(current, cmd, latestNodeWithId);
+    }
 
-    removeQueryPlansOfTerminated(compacted, terminatedQueries);
+    final List<QueuedCommand> compacted = new LinkedList<>();
+    while (current != null) {
+      // traverse backwards and add each next node to the start of the list
+      compact(current).ifPresent(cmd -> compacted.add(0, cmd));
+      current = current.prev;
+    }
 
     return compacted;
   }
 
-  private static Set<QueuedCommand> findTerminatedQueriesAndRemoveTerminateCommands(
-      final List<QueuedCommand> commands
-  ) {
-    final Map<QueryId, QueuedCommand> queries = new HashMap<>();
-    final Set<QueuedCommand> terminatedQueries = Collections.newSetFromMap(new IdentityHashMap<>());
-
-    final Iterator<QueuedCommand> it = commands.iterator();
-    while (it.hasNext()) {
-      final QueuedCommand cmd = it.next();
-
-      // Find known queries:
-      final Command command =
-          cmd.getAndDeserializeCommand(InternalTopicSerdes.deserializer(Command.class));
-      if (command.getPlan().isPresent()
-          && command.getPlan().get().getQueryPlan().isPresent()
-      ) {
-        final QueryId queryId =
-            command.getPlan().get().getQueryPlan().get().getQueryId();
-        queries.putIfAbsent(queryId, cmd);
+  private static final class CompactedNode {
+
+    final CompactedNode prev;
+    final QueuedCommand queued;
+    final Command command;
+
+    boolean shouldSkip = false;
+
+    public static CompactedNode of(
+        final CompactedNode prev,
+        final QueuedCommand queued,
+        final Map<QueryId, CompactedNode> latestNodeWithId
+    ) {
+      final Command command = queued.getAndDeserializeCommand(
+          InternalTopicSerdes.deserializer(Command.class)
+      );
+
+      final Optional<KsqlPlan> plan = command.getPlan();
+      if (queued.getAndDeserializeCommandId().getType() == Type.TERMINATE) {
+        final QueryId queryId = new QueryId(queued.getAndDeserializeCommandId().getEntity());
+        markShouldSkip(queryId, latestNodeWithId);
+
+        // terminate commands don't get added to the list of commands to execute
+        // because we "execute" them in this class by removing query plans from
+        // terminated queries
+        return prev;
+      } else if (!plan.isPresent() || !plan.get().getQueryPlan().isPresent()) {
+        // DDL
```

Comment:
consider asserting that if there is a previous stream/table with this ID, that it doesn't have a query associated with it?

Reply:
I don't know how that would be possible? DDL commands don't have any associated queryID, so even though there might be sources with the same name previously, they wouldn't be the same query